bouncer/shared/utils/indexer.ts

```diff
@@ -0,0 +1,101 @@
+import { z } from 'zod';
+import { sleep } from 'shared/utils';
+import prisma from './prisma_client';
+
+export type EventName = `${string}.${string}` | `.${string}`;
+type EventTime = {
+  startFromBlock: number;
+  endBeforeBlock?: number;
+};
+
+// ------------ Types for choosing an event of multiple alternatives   ---------------
+
+export type EventDescription = { name: EventName; schema: z.ZodTypeAny };
+
+export type EventDescriptions = Record<string, EventDescription>;
+
+export type ChooseSingleEvent<S extends Record<string, EventDescription>> = {
+  [K in keyof S]: {
+    key: K;
+    data: z.infer<S[K]['schema']>;
+    blockHeight: number;
+  };
+}[keyof S];
+
+// ------------ Querying the indexer database --------------
+export const findOneEventOfMany = async <Descriptions extends EventDescriptions>(
+  descriptions: Descriptions,
+  timing: EventTime,
+): Promise<ChooseSingleEvent<Descriptions>> => {
+  let foundEventsKeyAndData: { key: string; data: unknown; blockHeight: number }[] = [];
+  while (foundEventsKeyAndData.length === 0) {
+    const matchingEvents = await prisma.event.findMany({
+      where: {
+        OR: Object.values(descriptions).map((d) => ({
+          name: d.name.startsWith('.') ? { endsWith: d.name } : { equals: d.name },
+        })),
+        block: {
+          height: {
+            gte: timing.startFromBlock,
+            lt: timing.endBeforeBlock,
+          },
+        },
+      },
+      include: {
+        block: true,
+      },
+    });
+
+    if (matchingEvents) {
+      foundEventsKeyAndData = matchingEvents.flatMap((event) => {
+        const schemas = Object.entries(descriptions).flatMap(([key, d]) =>
+          event.name.includes(d.name) ? [{ key, schema: d.schema }] : [],
+        );
+        if (schemas.length === 0) {
+          throw new Error(
+            `Unexpected internal error in 'findOneOfMany': there where no event descriptions found matching the chosen event ${JSON.stringify(event)}. The database query might be off.`,
+          );
+        }
+
+        // Even though we found all events that match the given names, we have to check whether they
+        // also match the given schema.
+        const parsingResults = schemas.flatMap(({ key, schema }) => {
+          const r = schema.safeParse(event.args);
+          return r.success ? [{ key, data: r.data, blockHeight: event.block.height }] : [];
```

Comment:
It might be worth logging something here (?) - in general don't we expect the schema to match? Or does 'schema' also imply 'value' rather than just shape?